utilities/blob_db/blob_compaction_filter.cc

```diff
@@ -13,57 +13,128 @@
+    if (ucf == nullptr) {
+      return Decision::kKeep;
+    }
+    // Apply user compaction filter for inlined data.
+    decision = ucf->FilterV2(level, key, value_type, value, &new_filter_value,
+                             skip_until);
+  } else {
+    BlobIndex blob_index;
+    Status s = blob_index.DecodeFrom(value);
+    if (!s.ok()) {
+      // Unable to decode blob index. Keeping the value.
+      return Decision::kKeep;
+    }
+    if (blob_index.HasTTL() && blob_index.expiration() <= current_time_) {
+      // Expired
+      expired_count_++;
+      expired_size_ += key.size() + value.size();
+      return Decision::kRemove;
+    }
+    if (!blob_index.IsInlined() &&
+        blob_index.file_number() < context_.next_file_number &&
+        context_.current_blob_files.count(blob_index.file_number()) == 0) {
+      // Corresponding blob file gone (most likely, evicted by FIFO eviction).
       evicted_count_++;
       evicted_size_ += key.size() + value.size();
       return Decision::kRemove;
     }
+    if (context_.fifo_eviction_seq > 0 && blob_index.HasTTL() &&
+        blob_index.expiration() < context_.evict_expiration_up_to) {
+      // Hack: Internal key is passed to BlobIndexCompactionFilter for it to
+      // get sequence number.
+      key_parsed = ParseInternalKey(key, &ikey);
+      assert(key_parsed);
+      // Remove keys that could have been remove by last FIFO eviction.
+      // If get error while parsing key, ignore and continue.
+      if (key_parsed && ikey.sequence < context_.fifo_eviction_seq) {
+        evicted_count_++;
+        evicted_size_ += key.size() + value.size();
+        return Decision::kRemove;
+      }
+    }
+    // Apply user compaction filter for all non-TTL blob data.
+    if (ucf != nullptr && !blob_index.HasTTL()) {
+      // Hack: Internal key is passed to BlobIndexCompactionFilter for it to
+      // get sequence number.
+      key_parsed = ParseInternalKey(key, &ikey);
+      assert(key_parsed);
+      if (!key_parsed) {
+        return Decision::kKeep;
+      }
+      // Read value from blob file.
+      PinnableSlice blob;
+      CompressionType compression_type = kNoCompression;
+      if (!ReadBlobFromOldFile(ikey.user_key, blob_index, &blob, true,
+                               &compression_type)) {
+        // Keep the blob index in case fails to read the blob value.
+        return Decision::kKeep;
+      }
+      decision = ucf->FilterV2(level, ikey.user_key, value_type, blob,
+                               &new_filter_value, skip_until);
+    }
   }
-  return Decision::kKeep;
+  // If value changed, update value either inline or into blob file.
+  if (decision == Decision::kChangeValue) {
+    BlobDBImpl* const blob_db_impl = context_.blob_db_impl;
+    assert(blob_db_impl);
+    if (new_filter_value.size() > blob_db_impl->bdb_options_.min_blob_size) {
+      if (!OpenNewBlobFileIfNeeded()) {
+        return Decision::kKeep;
```

Comment:
How to handle I/O errors here is a good question; we don't want to end up with partially written/potentially corrupt blob files. Note that if we've already put some blobs in a new blob file and emitted the corresponding `BlobIndex`es, we cannot simply give up on the file (because then you would have dangling pointers). For GC, we simply abort the entire compaction in such cases. Will think about this more.